runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -2679,6 +2679,39 @@ bool isConditionCodeSetForCompare(TR::Node *node, bool *jumpOnOppositeCondition)
    return false;
    }
 
+void setImplicitNULLCHKExceptionInfo(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR_ASSERT(node->hasFoldedImplicitNULLCHK(), "Attempt to set exception info on BNDCHK without implicit NULLCHK");
+
+   TR::Compilation *comp = cg->comp();
+   TR::Instruction *faultingInstruction = cg->getImplicitExceptionPoint();
+   if (faultingInstruction)
+      {
+      // Check and correctly set the implicit exception point
+      //
+      // The compare analyzer may have generated a secondary load to
+      // resolved a compressed pointer and incorrectly set the exception point.
+      // If that is the case, correctly set the exception point on
+      // the comparison that actually throws the null pointer exception.
+      //
+      // The last instruction is a branch, the comparison is before.
+      TR::Instruction *cmp = cg->getAppendInstruction()->getPrev();
+      TR::InstOpCode::Mnemonic mnemonic = cmp->getOpCodeValue();
+      bool isComparisonMemForm = mnemonic == TR::InstOpCode::CMP4MemReg ||
+                                 mnemonic == TR::InstOpCode::CMP4RegMem;
+      if (comp->useCompressedPointers() && faultingInstruction != cmp && isComparisonMemForm)
+         {
+         faultingInstruction = cmp;
+         cg->setImplicitExceptionPoint(faultingInstruction);
```

Comment:
Maybe you could add a `traceMsg` (guarded with TraceCG) here indicating that the found faultingInstruction was changed.  This will make it clear in the log this happened.